pipeline/backend/kubernetes/kubernetes.go

```diff
@@ -36,37 +54,237 @@ func (e *kube) IsAvailable() bool {
-	// GET /api/v1/namespaces/{namespace}/pods/{name}/log
-	return nil, nil
+func (e *kube) Tail(ctx context.Context, step *types.Step) (io.ReadCloser, error) {
+	podName := podName(step)
+
+	up := make(chan bool)
+
+	podUpdated := func(old, new interface{}) {
+		pod := new.(*v1.Pod)
+		if pod.Name == podName {
+			switch pod.Status.Phase {
+			case v1.PodRunning, v1.PodSucceeded, v1.PodFailed:
+				up <- true
+			}
+		}
+	}
+
+	// TODO 5 seconds is against best practice, k3s didn't work otherwise
+	si := informers.NewSharedInformerFactoryWithOptions(e.client, 5*time.Second, informers.WithNamespace(e.namespace))
+	si.Core().V1().Pods().Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: podUpdated,
+		},
+	)
+	si.Start(wait.NeverStop)
+
+	<-up
+
+	opts := &v1.PodLogOptions{
+		Follow: true,
+	}
+
+	logs, err := e.client.CoreV1().RESTClient().Get().
+		Namespace(e.namespace).
+		Name(podName).
+		Resource("pods").
+		SubResource("log").
+		VersionedParams(opts, scheme.ParameterCodec).
+		Stream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rc, wc := io.Pipe()
+
+	go func() {
+		defer e.logs.Reset()
+		defer logs.Close()
+		defer wc.Close()
+		defer rc.Close()
+
+		systemLogs := io.NopCloser(bytes.NewReader(e.logs.Bytes()))
+		_, err := io.Copy(wc, systemLogs)
+		if err != nil {
+			return
+		}
+		_, err = io.Copy(wc, logs)
+		if err != nil {
+			return
+		}
+	}()
+	return rc, nil
+
+	// rc := io.NopCloser(bytes.NewReader(e.logs.Bytes()))
+	// e.logs.Reset()
+	// return rc, nil
 }
 
 // Destroy the pipeline environment.
-func (e *kube) Destroy(context.Context, *types.Config) error {
-	// DELETE /api/v1/namespaces/{name}
+func (e *kube) Destroy(ctx context.Context, conf *types.Config) error {
+	var gracePeriodSeconds int64 = 0 // immediately
+	dpb := metav1.DeletePropagationBackground
+
+	deleteOpts := metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriodSeconds,
+		PropagationPolicy:  &dpb,
+	}
+
+	for _, stage := range conf.Stages {
+		for _, step := range stage.Steps {
+			e.logs.WriteString("Deleting pod\n")
+			if err := e.client.CoreV1().Pods(e.namespace).Delete(ctx, podName(step), deleteOpts); err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, stage := range conf.Stages {
+		if stage.Alias == "services" {
+			for _, step := range stage.Steps {
+				e.logs.WriteString("Deleting service\n")
+				// TODO: support ports setting
+				// svc, err := Service(e.namespace, step.Name, step.Alias, step.Ports)
+				svc, err := Service(e.namespace, step.Name, step.Alias, []string{})
+				if err != nil {
+					return err
+				}
+				if err := e.client.CoreV1().Services(e.namespace).Delete(ctx, svc.Name, deleteOpts); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	for _, vol := range conf.Volumes {
+		pvc := PersistentVolumeClaim(e.namespace, vol.Name, e.storageClass, e.volumeSize, e.storageRwx)
+		err := e.client.CoreV1().PersistentVolumeClaims(e.namespace).Delete(ctx, pvc.Name, deleteOpts)
+		if err != nil {
```

Comment:
When canceling the job via the UI the ctx we receive is already canceled, and therefore the .Delete command errors and the containers don't get deleted.
And don't abort on NotFound errors.

Re-used the way around this from the docker backend, use a noContext instead

```suggestion
func (e *kube) Destroy(ctx context.Context, conf *types.Config) error {
	var gracePeriodSeconds int64 = 0 // immediately
	dpb := metav1.DeletePropagationBackground

	deleteOpts := metav1.DeleteOptions{
		GracePeriodSeconds: &gracePeriodSeconds,
		PropagationPolicy:  &dpb,
	}

	for _, stage := range conf.Stages {
		for _, step := range stage.Steps {
			e.logs.WriteString("Deleting pod\n")
			if err := e.client.CoreV1().Pods(e.namespace).Delete(noContext, podName(step), deleteOpts); err != nil && !errors.IsNotFound(err) {
				return err
			}
		}
	}

	for _, stage := range conf.Stages {
		if stage.Alias == "services" {
			for _, step := range stage.Steps {
				e.logs.WriteString("Deleting service\n")
				// TODO: support ports setting
				// svc, err := Service(e.namespace, step.Name, step.Alias, step.Ports)
				svc, err := Service(e.namespace, step.Name, step.Alias, []string{})
				if err != nil {
					return err
				}
				if err := e.client.CoreV1().Services(e.namespace).Delete(noContext, svc.Name, deleteOpts); err != nil && !errors.IsNotFound(err) {
					return err
				}
			}
		}
	}

	for _, vol := range conf.Volumes {
		pvc := PersistentVolumeClaim(e.namespace, vol.Name, e.storageClass, e.volumeSize, e.storageRwx)
		err := e.client.CoreV1().PersistentVolumeClaims(e.namespace).Delete(noContext, pvc.Name, deleteOpts)
		if err != nil && !errors.IsNotFound(err) {
```

Reply:
Good catch can you open a pull for current docker & add some descriptive code comments ❤️

Reply:
If errNotFound we should trace log it ... so we can detect missbehaviour

Reply:
> Good catch can you open a pull for current docker & add some descriptive code comments
The docker-backend already has this stuff, where I got the idea to use the noContext: https://github.com/woodpecker-ci/woodpecker/blob/master/pipeline/backend/docker/docker.go#L203

As for the code comments and trace log, I've created a PR to @anbraten k8s branch again rather than editing this. These changes got very messy in to do in the change suggestions.

https://github.com/anbraten/woodpecker/pull/4